galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker.go

```diff
@@ -0,0 +1,224 @@
+func (w Workload) PortNumber() uint32 {
+	return w.portNumber
+}
+
+// PortName is the name of the port used by the workload. If set (not ""),
+// then PortNumber must be the default value (0).
+func (w Workload) PortName() string {
+	return w.portName
+}
+
+// PolicyChecker allows callers to add a set of v1alpha1.Policy objects in
+// the mesh. Once these are loaded, you can query whether or not a specific
+// Workload will require MTLS when an incoming connection occurs using the
+// IsServiceMTLSEnforced() call.
+type PolicyChecker struct {
+	// meshHasStrictMTLSPolicy tracks whether or not mTLS is strictly enforced on the mesh.
+	meshHasStrictMTLSPolicy bool
+
+	namespaceToMTLSMode map[string]strictMTLSMode
+	workloadToMTLSMode  map[Workload]strictMTLSMode
+}
+
+// strictMTLSMode is a helper type used to represent whether or not MTLS was
+// explicitly enabled, explicitly disabled, or just not present. Useful for the
+// IsServiceMTLSEnforced check to determine which policy applies.
+type strictMTLSMode int
+
+const (
+	strictMTLSUnset strictMTLSMode = iota
+	strictMTLSExplicitlyEnabled
+	strictMTLSExplicitlyDisabled
+)
+
+// NewPolicyChecker creates a new PolicyChecker instance.
+func NewPolicyChecker() *PolicyChecker {
+	return &PolicyChecker{
+		namespaceToMTLSMode: make(map[string]strictMTLSMode),
+		workloadToMTLSMode:  make(map[Workload]strictMTLSMode),
+	}
+}
+
+// AddMeshPolicy adds a mesh-level policy to the checker. Note that there can
+// only be at most one mesh level policy in effect.
+func (pc *PolicyChecker) AddMeshPolicy(p *v1alpha1.Policy) {
+	pc.meshHasStrictMTLSPolicy = doesPolicyEnforceMTLS(p)
+}
+
+// AddPolicy adds a new policy object to the PolicyChecker to use when later
+// determining if a service is MTLS-enforced. The namespace of the policy is
+// also provided as some policies can target the local namespace.
+func (pc *PolicyChecker) AddPolicy(namespace string, p *v1alpha1.Policy) error {
+	var policyMode strictMTLSMode
+	if doesPolicyEnforceMTLS(p) {
+		policyMode = strictMTLSExplicitlyEnabled
+	} else {
+		policyMode = strictMTLSExplicitlyDisabled
+	}
+
+	if len(p.Targets) == 0 {
+		// Rule targets the namespace.
+		pc.namespaceToMTLSMode[namespace] = policyMode
+		return nil
+	}
+	// Discover the targeted workload and take note. Should normalize.
+	for _, target := range p.Targets {
+		fqdn := util.ConvertHostToFQDN(namespace, target.Name)
+
+		if len(target.Ports) == 0 {
+			// Policy targets all ports on workload
+			pc.workloadToMTLSMode[NewWorkload(fqdn)] = policyMode
+		}
+
+		for _, port := range target.Ports {
+			if port.GetName() != "" {
+				pc.workloadToMTLSMode[NewWorkloadWithPortName(fqdn, port.GetName())] = policyMode
+			} else if port.GetNumber() != 0 {
+				pc.workloadToMTLSMode[NewWorkloadWithPortNumber(fqdn, port.GetNumber())] = policyMode
+			} else {
+				// Unhandled case!
+				return fmt.Errorf("policy has a port with no name/number for target %s", target.Name)
```

Comment:
Does Galley allow this case? If Galley doesn't reject this, probably we need an analyzer for this case.

Reply:
I think the protocol buffer implementation validates this via a oneof constraint:

https://github.com/istio/api/blob/master/authentication/v1alpha1/policy.proto#L411

I added the unhandled case to throw an error if we ever allow specifying a port via some other mechanism.